// squarelink/dist/iframe.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _config = require('./config');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Iframe = function () {
  /**
   * @param {string} url
   */
  function Iframe(url, params) {
    (0, _classCallCheck3.default)(this, Iframe);

    this.url = url;
    this.params = params;
    this.open = true;
    this._createIframe();
    this._addCloseListeners();
    this._addMessageListeners();
  }

  (0, _createClass3.default)(Iframe, [{
    key: 'close',
    value: function close() {
      if (!this.open) return;
      this.container.parentNode.removeChild(this.container);
      if (this.onClosed) this.onClosed(this.error);
      this.open = false;
    }
  }, {
    key: '_addCloseListeners',
    value: function _addCloseListeners() {
      var _this = this;

      this.container.addEventListener('click', function () {
        _this.close();
      });
      var self = this;
      document.onkeydown = function (evt) {
        evt = evt || window.event;
        if (evt.keyCode == 27 && self.open) {
          self.close();
        }
      };
    }
  }, {
    key: '_addMessageListeners',
    value: function _addMessageListeners() {
      var _this2 = this;

      var self = this;
      var paramsSent = false;
      window.addEventListener('message', function (e) {
        var _e$data = e.data,
            origin = _e$data.origin,
            height = _e$data.height,
            type = _e$data.type,
            error = _e$data.error;

        if (origin === 'squarelink-iframe') {
          if (type === 'resize') {
            self.iframe.style = styles.iframe(height + 'px', 'none');
            return;
          } else if (type === 'error') {
            self.error = error;
            self.close();
          } else if (type === 'onload' && !paramsSent) {
            paramsSent = true;
            _this2.iframe.contentWindow.postMessage({
              origin: 'squarelink-web3-sdk',
              params: _this2.params
            }, '*');
          }
        }
      }, false);
    }
  }, {
    key: '_createIframe',
    value: function _createIframe() {
      /* INITIALIZE IFRAME CONTAINER */
      var container = document.createElement('div');
      container.id = 'squarelink-iframe-container';
      container.style = styles.container;

      /* INITIALIZE PRELOADER */
      var preloader = '<div class="squarelink-preloader" id="squarelink-preloader"><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div><div></div></div>';
      container.innerHTML = preloader;

      /* INITIALIZE IFRAME */
      var iframe = document.createElement('iframe');
      iframe.src = _config.IFRAME_URL + '/?url=' + encodeURI(this.url);
      iframe.id = 'squarelink-iframe';
      iframe.style = styles.iframe();
      iframe.onload = function () {
        var pl = document.getElementById('squarelink-preloader');
        pl.parentNode.removeChild(pl);
      };
      container.appendChild(iframe);
      this.iframe = iframe;
      this.container = container;

      /* LOAD IFRAME CONTAINER */
      document.body.appendChild(container);
    }
  }]);
  return Iframe;
}();

exports.default = Iframe;


var styles = {
  iframe: function iframe() {
    var height = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '200px';
    var border = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '3px solid #fff';

    return '\n      position: absolute;\n      height: ' + height + ';\n      width: 360px;\n      top: 50%;\n      left: 50%;\n      transform:\n      translate(-50%, -50%);\n      border: 0px transparent;\n      border-radius: 10px;\n      -webkit-border-radius: 10px;\n      -moz-border-radius: 10px;\n      border-radius: 10px;\n      -khtml-border-radius: 10px;\n      border: ' + border + ';\n      z-index: 2147483647;\n      box-shadow: 0 10px 30px 4px rgba(0,0,0,.33);\n      background: none;\n    ';
  },
  container: '\n    position: fixed;\n    height: 100%;\n    width: 100%;\n    left: 0;\n    right: 0;\n    top: 0;\n    bottom: 0;\n    z-index: 2147483647;\n    background: rgba(0,0,0,0.5);\n  '
};
module.exports = exports.default;
